File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        from collections import defaultdict
        dic = defaultdict(int)
        l = 0
        maxx = 0
        for r ,ch in  enumerate(s):

            dic[ch] += 1
            while dic[ch] > 1:
                dic[s[l]] -= 1
                if dic[s[l]] == 0:
                    dic.pop(s[l])
                l+=1
            maxx = max(maxx,r-l+1)
        
        return maxx
            

        

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        int left = 0;
        HashMap<Character, Integer> count = new HashMap<>();
        int maxx = 0;
        for(int r = 0; r<s.length(); r++){
            char c = s.charAt(r);
            count.put(c , count.getOrDefault(c,0)+1);

            while (count.get(c) >1 && r>left){
                char lc = s.charAt(left);
                count.put(lc, count.getOrDefault(lc, 0)- 1);
                left++;
            }
            maxx = Math.max(maxx,r-left+1);



        }
        return maxx;
        
    }
}